bundles/org.eclipse.jface.text/src/org/eclipse/jface/text/DefaultDocumentAdapter.java

```diff
@@ -220,7 +221,26 @@ public String getTextRange(int offset, int length) {
 	@Override
 	public void replaceTextRange(int pos, int length, String text) {
 		try {
-			fDocument.replace(pos, length, text);
+			String[][] bracketTypes= {
+					{ SpecialCharacterConstants.openParentheses, SpecialCharacterConstants.openSquareBracket, SpecialCharacterConstants.openCurlyBracket, SpecialCharacterConstants.openAngleBracket,
+							SpecialCharacterConstants.singleQuote, SpecialCharacterConstants.doubleQuotes },
+					{ SpecialCharacterConstants.closeParentheses, SpecialCharacterConstants.closeSquareBracket, SpecialCharacterConstants.closeCurlyBracket,
+							SpecialCharacterConstants.closeAngleBracket, SpecialCharacterConstants.singleQuote, SpecialCharacterConstants.doubleQuotes }
+			};
+			boolean containsSpecialBrackets= Arrays.asList(bracketTypes[0]).contains(text);
```

Comment:
You can store this List as constant "private final static", please. Hint: better use` List.of(...) `